dace/transformation/interstate/loop_to_map.py

```diff
@@ -0,0 +1,114 @@
+import sympy as sp
+import networkx as nx
+from typing import List, Optional, Tuple
+
+from dace import dtypes, memlet, nodes, registry, sdfg as sd, symbolic, subsets
+from dace.properties import Property, make_properties, CodeBlock
+from dace.sdfg import graph as gr, nodes
+from dace.sdfg import utils as sdutil
+from dace.frontend.python.astutils import ASTFindReplace
+from dace.transformation.interstate.loop_detection import (DetectLoop,
+                                                           find_for_loop)
+import dace.transformation.helpers as helpers
+
+
+@registry.autoregister
+class LoopToMap(DetectLoop):
+    """Convert a control flow loop into a dataflow map. Currently only supports
+       the simple case where there is no overlap between inputs and outputs in
+       the body of the loop, and where the loop body only consists of a single
+       state.
+    """
+
+    @staticmethod
+    def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
+        # Is this even a loop
+        if not DetectLoop.can_be_applied(graph, candidate, expr_index, sdfg,
+                                         strict):
+            return False
+
+        guard = graph.node(candidate[DetectLoop._loop_guard])
+        begin = graph.node(candidate[DetectLoop._loop_begin])
+
+        # Guard state should contain any dataflow
+        if len(guard.nodes()) != 0:
+            return False
+
+        # Only support loops with a single-state body
+        begin_outedges = graph.out_edges(begin)
+        if len(begin_outedges) != 1 or begin_outedges[0].dst != guard:
+            return False
+
+        # If loop cannot be detected, fail
+        _, rng = find_for_loop(graph, guard, begin)
+        if not rng:
+            return False
+
+        # Currently only detect the trivial case where the set of containers
+        # that are read are completely disjoint from those that are written
+        read_set, write_set = helpers.read_and_write_set(begin)
+        if len(read_set & write_set) != 0:
+            return False
+
+        # TODO: Detect that the iteration variable isn't used anywhere else,
+        #       or set it to the final value of the map iterator.
+
+        return True
+
+    def apply(self, sdfg):
+        # Obtain loop information
+        guard: sd.SDFGState = sdfg.node(self.subgraph[DetectLoop._loop_guard])
+        body: sd.SDFGState = sdfg.node(self.subgraph[DetectLoop._loop_begin])
+        after: sd.SDFGState = sdfg.node(self.subgraph[DetectLoop._exit_state])
+
+        # Obtain iteration variable, range, and stride
+        itervar, rng = find_for_loop(sdfg, guard, body)
+
+        source_nodes = body.source_nodes()
+        sink_nodes = body.sink_nodes()
+
+        map = nodes.Map(body.label + "_map", [itervar], [rng])
+        entry = nodes.MapEntry(map)
+        exit = nodes.MapExit(map)
+        body.add_node(entry)
+        body.add_node(exit)
+
+        # Reroute all memlets through the entry and exit nodes
```

Comment:
what if there is a free tasklet there? in that case you should "route" it as `body.add_edge(entry, None, tasklet, None, memlet.Memlet())`

Reply:
Implemented in a941fad1